### ISSUE TITLE
Cannot compile on Arch Linux with assimp

### ISSUE THREAD

**Issue by Rheinwalt** [isl-org/Open3D#6140]
### Checklist

- [X] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](http://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](http://www.open3d.org/docs/release/) and the [latest documentation](http://www.open3d.org/docs/latest/) (for `master` branch).

#### I first cloned Open3D by:
```
git clone https://github.com/isl-org/Open3D.git
cd Open3D
```

#### Then, I build Open3D (on Arch Linux without CUDA) with:
```
mkdir build
cd build
cmake -DBUILD_SHARED_LIBS=OFF -DBUILD_CUDA_MODULE=OFF -DUSE_BLAS=ON -DBUILD_WEBRTC=OFF ..
make -j6
```

I have assimp installed, even pyassimp, but Open3D wants to use their own third-party version. Which is fine, but it fails to build?

### Error message

```shell
In static member function ‘static _Up* std::__copy_move<_IsMove, true, std::random_access_iterator_tag>::__copy_m(_Tp*, _Tp*, _Up*) [with _Tp = long unsigned int; _Up = long unsigned int; bool _IsMove = false]’,
    inlined from ‘_OI std::__copy_move_a2(_II, _II, _OI) [with bool _IsMove = false; _II = long unsigned int*; _OI = long unsigned int*]’ at /usr/include/c++/13.1.1/bits/stl_algobase.h:506:30,
    inlined from ‘_OI std::__copy_move_a1(_II, _II, _OI) [with bool _IsMove = false; _II = long unsigned int*; _OI = long unsigned int*]’ at /usr/include/c++/13.1.1/bits/stl_algobase.h:533:42,
    inlined from ‘_OI std::__copy_move_a(_II, _II, _OI) [with bool _IsMove = false; _II = long unsigned int*; _OI = long unsigned int*]’ at /usr/include/c++/13.1.1/bits/stl_algobase.h:540:31,
    inlined from ‘_OI std::copy(_II, _II, _OI) [with _II = long unsigned int*; _OI = long unsigned int*]’ at /usr/include/c++/13.1.1/bits/stl_algobase.h:633:7,
    inlined from ‘std::vector<bool, _Alloc>::iterator std::vector<bool, _Alloc>::_M_copy_aligned(const_iterator, const_iterator, iterator) [with _Alloc = std::allocator<bool>]’ at /usr/include/c++/13.1.1/bits/stl_bvector.h:1305:28,
    inlined from ‘void std::vector<bool, _Alloc>::_M_reallocate(size_type) [with _Alloc = std::allocator<bool>]’ at /usr/include/c++/13.1.1/bits/vector.tcc:851:40,
    inlined from ‘void std::vector<bool, _Alloc>::reserve(size_type) [with _Alloc = std::allocator<bool>]’ at /usr/include/c++/13.1.1/bits/stl_bvector.h:1093:17,
    inlined from ‘Assimp::MDL::IntSharedData_MDL7::IntSharedData_MDL7()’ at /home/a/src/Open3D/build/assimp/src/ext_assimp/code/AssetLib/MDL/MDLFileData.h:879:32,
    inlined from ‘void Assimp::MDLImporter::InternReadFile_3DGS_MDL7()’ at /home/a/src/Open3D/build/assimp/src/ext_assimp/code/AssetLib/MDL/MDLLoader.cpp:1351:29:
/usr/include/c++/13.1.1/bits/stl_algobase.h:437:30: error: ‘void* __builtin_memmove(void*, const void*, long unsigned int)’ forming offset 8 is out of the bounds [0, 8] [-Werror=array-bounds=]
  437 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[5]: *** [code/CMakeFiles/assimp.dir/build.make:1476: code/CMakeFiles/assimp.dir/AssetLib/MDL/MDLLoader.cpp.o] Error 1
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [CMakeFiles/Makefile2:169: code/CMakeFiles/assimp.dir/all] Error 2
make[3]: *** [Makefile:136: all] Error 2
make[2]: *** [CMakeFiles/ext_assimp.dir/build.make:86: assimp/src/ext_assimp-stamp/ext_assimp-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:979: CMakeFiles/ext_assimp.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

### Open3D, Python and System information

```markdown
- Operating system: Arch Linux
- Python version: Python 3.11
- Open3D version: current git
- System architecture: x86
- Is this a remote workstation?: no
- How did you install Open3D?: build from source
- Compiler version (if built from source): gcc 13.1.1 / clang 15.0.7
```

**Comment by Rheinwalt** [comment]
If I enforce cmake to use the system assimp it builds that, but fails at tbb:

```
cd build
cmake -DUSE_SYSTEM_ASSIMP=ON ..
make
```

Will produce:

```
[ 21%] Built target ext_vtk
[ 21%] Built target ext_directxheaders
[ 22%] Built target ext_directxmath
[ 23%] Built target ext_uvatlas
[ 23%] Performing build step for 'ext_tbb'
[  2%] Building CXX object CMakeFiles/tbb_static.dir/src/old/concurrent_vector_v2.cpp.o
In file included from /home/a/src/Open3D/build/tbb/src/ext_tbb/src/old/concurrent_vector_v2.cpp:20:
/home/a/src/Open3D/build/tbb/src/ext_tbb/include/tbb/task.h:300:20: error: declaration of ‘tbb::task& tbb::internal::task_prefix::task()’ changes meaning of ‘task’ [-Wchanges-meaning]
  300 |         tbb::task& task() {return *reinterpret_cast<tbb::task*>(this+1);}
      |                    ^~~~
/home/a/src/Open3D/build/tbb/src/ext_tbb/include/tbb/task.h:252:9: note: used here to mean ‘class tbb::task’
  252 |         task* next_offloaded;
      |         ^~~~
/home/a/src/Open3D/build/tbb/src/ext_tbb/include/tbb/task.h:43:7: note: declared here
   43 | class task;
      |       ^~~~
make[5]: *** [CMakeFiles/tbb_static.dir/build.make:90: CMakeFiles/tbb_static.dir/src/old/concurrent_vector_v2.cpp.o] Error 1
make[4]: *** [CMakeFiles/Makefile2:114: CMakeFiles/tbb_static.dir/all] Error 2
make[3]: *** [Makefile:136: all] Error 2
make[2]: *** [CMakeFiles/ext_tbb.dir/build.make:86: tbb/src/ext_tbb-stamp/ext_tbb-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:1762: CMakeFiles/ext_tbb.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

**Comment by Rheinwalt** [comment]
Should I reopen this with a new issue?

**Comment by MoRoBe-Work** [comment]
@Rheinwalt did you ultimately solve this? I did not find a new issue you opened. For me it also fails on building tbb and being more used to Python I have some trouble understanding enough of the build process to solve this on my own in a timely manner. Any hints? Error (see below) is pretty much the same with the slight difference that I'm on Manjaro using Pamac for building the package, so the paths are different. But I could switch to manually build it if this allowed me to make necessary changes.

Error:
`In file included from /var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/src/old/concurrent_vector_v2.cpp:20:
/var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/include/tbb/task.h:300:20: error: declaration of ‘tbb::task& tbb::internal::task_prefix::task()’ changes meaning of ‘task’ [-Wchanges-meaning]
  300 |         tbb::task& task() {return *reinterpret_cast<tbb::task*>(this+1);}
      |                    ^~~~
/var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/include/tbb/task.h:252:9: note: used here to mean ‘class tbb::task’
  252 |         task* next_offloaded;
      |         ^~~~
/var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/include/tbb/task.h:43:7: note: declared here
   43 | class task;
      |       ^~~~`

**Comment by rogeroberholtzer** [comment]
I have the error reported in the initial message. Is there a solution?

- Open3D version: 0.17.0 (current GIT master)
- Operating system: openSUSE Tumbleweed 20231210
- Python version: Python 3.11
- System architecture: x86_64

**Comment by greenlet** [comment]
> @Rheinwalt did you ultimately solve this? I did not find a new issue you opened. For me it also fails on building tbb and being more used to Python I have some trouble understanding enough of the build process to solve this on my own in a timely manner. Any hints? Error (see below) is pretty much the same with the slight difference that I'm on Manjaro using Pamac for building the package, so the paths are different. But I could switch to manually build it if this allowed me to make necessary changes.
> 
> Error: `In file included from /var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/src/old/concurrent_vector_v2.cpp:20: /var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/include/tbb/task.h:300:20: error: declaration of ‘tbb::task& tbb::internal::task_prefix::task()’ changes meaning of ‘task’ [-Wchanges-meaning] 300 | tbb::task& task() {return *reinterpret_cast<tbb::task*>(this+1);} | ^~~~ /var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/include/tbb/task.h:252:9: note: used here to mean ‘class tbb::task’ 252 | task* next_offloaded; | ^~~~ /var/tmp/pamac-build-mbeeking/open3d-git/src/open3d-git/build/tbb/src/ext_tbb/include/tbb/task.h:43:7: note: declared here 43 | class task; | ^~~~`

Have you managed to overcome this? I have the same issue with forced system assimp but tbb build failed

**Comment by danrauch** [comment]
I ran into the same issue on Manjaro. My guess is that Open3D is only meant to be built on Ubuntu 18.04 / 20.04 (which use gcc verions >=9) and simply does not compile on gcc 13, due to some new warnings which get promoted to errors in the CMake config. It apparently built until gcc12, as the AUR package references it in its PKGBUILD: https://aur.archlinux.org/packages/open3d

Most (but not all) such warnings come from ancient 3rd party dependencies like oneDPL 20190522.

So either you modify the CMake files so that the warnings are ignored or you use a dev container / virtual machine with Ubuntu 20.04 (this is what I ended up doing).
